### ISSUE TITLE
Regression on 0.5.2

### ISSUE THREAD

**Issue by kamaradclimber** [CocoaPods/Molinillo#52]
Hello,

when using 0.5.2 I get an exception when using `berks update` command.
The issue appeared when starting to use 0.5.2.

Backtrace:

```
NoMethodError: undefined method `constraint' for nil:NilClass
  /home/g_seux/.gem/ruby/2.3.0/gems/solve-3.0.1/lib/solve/ruby_solver.rb:167:in `requirement_satisfied_by?'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/delegates/specification_provider.rb:24:in `block in requirement_satisfied_by?'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/delegates/specification_provider.rb:69:in `with_no_such_dependency_error_handling'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/delegates/specification_provider.rb:23:in `requirement_satisfied_by?'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolution.rb:345:in `block in attempt_to_swap_possibility'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolution.rb:345:in `each'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolution.rb:345:in `all?'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolution.rb:345:in `attempt_to_swap_possibility'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolution.rb:330:in `attempt_to_activate_existing_spec'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolution.rb:315:in `attempt_to_activate'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolution.rb:137:in `process_topmost_state'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolution.rb:78:in `resolve'
  /home/g_seux/.gem/ruby/2.3.0/bundler/gems/Molinillo-8d7454fba6d1/lib/molinillo/resolver.rb:42:in `resolve'
  /home/g_seux/.gem/ruby/2.3.0/gems/solve-3.0.1/lib/solve/ruby_solver.rb:200:in `resolve_with_error_wrapping'
  /home/g_seux/.gem/ruby/2.3.0/gems/solve-3.0.1/lib/solve/ruby_solver.rb:75:in `resolve'
  /home/g_seux/.gem/ruby/2.3.0/gems/solve-3.0.1/lib/solve.rb:64:in `it!'
  /home/g_seux/.gem/ruby/2.3.0/gems/berkshelf-5.1.0/lib/berkshelf/resolver.rb:78:in `resolve'
  /home/g_seux/.gem/ruby/2.3.0/gems/berkshelf-5.1.0/lib/berkshelf/installer.rb:175:in `install_from_universe'
  /home/g_seux/.gem/ruby/2.3.0/gems/berkshelf-5.1.0/lib/berkshelf/installer.rb:39:in `run'
  /home/g_seux/.gem/ruby/2.3.0/gems/berkshelf-5.1.0/lib/berkshelf/berksfile.rb:422:in `install'
  /home/g_seux/.gem/ruby/2.3.0/gems/berkshelf-5.1.0/lib/berkshelf/berksfile.rb:447:in `update'
```

I've tried to use the patch proposed on #49 without success.

**Comment by segiddins** [comment]
Is there a way to reproduce this in a Molinillo test case? Thanks!

**Comment by kamaradclimber** [comment]
thanks for your quick answer!

I don't think I will be able to provide a test case since berks is used to update ~200 internal cookbooks.

I can try to understand what leads to the issue to create a test case but it will take some time since I'm unfamiliar to this code base.

**Comment by segiddins** [comment]
Sure, I just don't see how I can understand the issue otherwise

**Comment by kamaradclimber** [comment]
First investigation leads to an edge of the dependency graph whose requirement is nil.
It is never set nil at creation but rather modified later (https://github.com/CocoaPods/Molinillo/blob/master/lib/molinillo/resolution.rb#L376)

I'll try to understand why

**Comment by kamaradclimber** [comment]
@segiddins I don't think I'll be able to provide a test case reproducing the error. I hope you could take it from here. Otherwise let's wait if other users came up with a reproducible case.

**Comment by segiddins** [comment]
Is it possible this was fixed in 0.5.3? Otherwise, I'm going to have to close this issue as unable to reproduce

**Comment by kamaradclimber** [comment]
The issue does not appear again at that time. I'll close the issue and reopen if we have another occurence
